feat: process xml download

eppdot opened the pull request:
## Description

We added the button to the diagram itself and also enhanced the filename that is used to store the download based on
`processName-v-version`. The download with enhanced names is shown in the process instance view, the process list view, the migration source & target views.

### Notes

- fix formatting
- refactor method to avoid any type param
- add DownloadBPMNDefinitionXML button
- update snapshots
- revert screenshot expectated threshold to default 0.2 -- make it more lax

## Related issues

closes #22005

eppdot committed:
feat: add download xml button to Diagram component

eppdot committed:
feat: enhance the diagram download file name with process name and version

eppdot commented on operate/client/src/modules/utils/processDefinition.ts:
```diff
@@ -0,0 +1,24 @@
+const getFullyQualifiedProcessDefinitionNameBy = (definition: any) => {
+  if(!definition) return undefined;
+
+  // for now the definition can be of different types, so we try to get the name and version from different properties
```
this is a bit ugly and we could split this also into 2-3 methods with explicit types or do some typecheck here. But we thought being pragmatic is good enough here 😅

vsgoulart reviewed (dismissed):
Please fix the formatting

I think should encapsulate the button logic into a single component called `<DownloadBPMNDefinitionXML processDefinitionKey={processDefinitionKey} />`

And then just consume the Tanstack query hook, it handle the caching for you and it won't make any extra calls and you can use a lazy query which will only execute query on click

konstantinos-mylosis committed:
feat: enhance the diagram download

fix formatting
refactor method to avoid any type param
add DownloadBPMNDefinitionXML button
update snapshots
revert screenshot expectated threshold to default 0.2

konstantinos-mylosis commented:
fix: 1st iteration of CI issues

TODO
> I think should encapsulate the button logic into a single component called `<DownloadBPMNDefinitionXML processDefinitionKey={processDefinitionKey} />`
> 
> And then just consume the Tanstack query hook, it handle the caching for you and it won't make any extra calls and you can use a lazy query which will only execute query on click

abremard reviewed (commented):
@konstantinos-mylosis there are still a couple of ts issues, let me know if you need help fixing them

abremard reviewed (commented):
Overall the updates look good to me, nice one! 
I have a couple of small comments/suggestions before approving

eppdot commented on operate/client/src/modules/components/DownloadBPMNDefinitionXML/index.tsx:
```diff
@@ -0,0 +1,97 @@
+      const link = document.createElement('a');
+      link.href = url;
+
+      link.download = `${_.kebabCase(
+        processDefinition
+          ? getDiagramNameByProcessDefinition(processDefinition)
+          : 'diagram',
+      )}.bpmn`;
```
I like the proposal and also think we could adjust the naming pattern to `a-b-c_v1.xml` instead of `a-b-c-v-1.xml` due to the usage on kebabCase on the full "name". Maybe just move kebabcase into the utilit function?

konstantinos-mylosis committed:
feat: simplify XML download button

move process definition name to util

abremard reviewed (approved):
Changes look good to me, I saw you just pushed a fix for the CI tests, I'm pre-approving

vsgoulart committed:
refactor: remove unused V1 diagram implementation

vsgoulart committed:
refactor: Do not rename download button

vsgoulart committed:
refactor: remove unused prop

vsgoulart committed:
refactor: make processDefinitionKey mandatory

vsgoulart committed:
refactor: Use lodash named import

vsgoulart committed:
refactor: rename diagram file name util

vsgoulart committed:
refactor: make passing an undefined definition impossible

vsgoulart committed:
refactor: lazy load download data

vsgoulart committed:
refactor: use global query mapping

vsgoulart committed:
test: remove wrong casting

vsgoulart committed:
test: remove wrong mocks

vsgoulart committed:
test: add test for download button conditional rendering
